### PR TITLE
fix: add y param  and fix n_features_in_ for sklearn compatibility

### LINKED ISSUE #3139: [BUG] sklearn `check_n_features_in_after_fitting` fail for causal discovery estimators
**Describe the bug**
If we run `check_estimator` on any causal discovery algorithm, two of sklearn's `check_estimator` tests fail: 

1. `check_fit_score_takes_y`: This fails as sklearn expects `score(X, y)` signature for scoring methods. This is not valid for causal discovery algorithms as we either compare learnt and true graphs or compare the learned graph to the data. 
2. `check_n_features_in_after_fitting`: This checks for the value of `n_features_in_` after fit and score methods. This check passes for the `fit` method but fails for our `score` method. It is unclear why it fails for the `score` method.

**To Reproduce**
```python
from sklearn.utils.estimator_checks import check_estimator
from pgmpy.causal_discovery import PC

check_estimator(PC(return_type="dag"))
```

**Expected behavior**
The `check_n_features_in_after_fitting` should pass for both `fit` and `score` methods.

**Additional context**
found this while working on #2839 and implementing the compliance checker in #3127. currently these are listed as expected failures in EXPECTED_FAILED_SKLEARN_CHECKS in `pgmpy/tests/test_causal_discovery/check_causal_discovery.py`

**Versions**
- pgmpy 1.0.0
- scikit-learn 1.8.0

### PR DESCRIPTION
**Pull request by Famous077**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [X] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [X] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [X] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [X] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md#ai-usage-policy)?
- [X] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.  

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.
I'm beginner in this org. So, I most used AI for understand the codebase easily with simple analogy, and also for understand issue. what type of issue is this , what are the problems that are occuring due to this issue.

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?  

I used pytest module to verify that the changes correctly address the issue. I haven't faced any edge cases situation. Haven't verify anything else than running tests.

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

No, I didn't generate any AI test. I used pytest module to verify after changes everything is working correctly or not.

### Issue number(s) that this pull request fixes
- Fixes #3139 

### List of changes to the codebase in this pull request
- Added a y=None parameter and self.n_features_in_ = X.shape[1] in causal_discovery/_base.py.

### CODE REVIEW

File: pgmpy/tests/test_causal_discovery/test_HillClimbSearch.py [hunk 1]
```diff
@@ -18,8 +18,8 @@
 
 def expected_failed_checks(estimator):
     return {
-        "check_fit_score_takes_y": "Causal discovery estimators do not take y parameter in score method.",
-        "check_n_features_in_after_fitting": "Failing for score method (not for fit) for unknown reason.",
+        # "check_fit_score_takes_y": "Causal discovery estimators do not take y parameter in score method.",
+        # "check_n_features_in_after_fitting": "Failing for score method (not for fit) for unknown reason.",
```
Comment (DARHWOLF) [inline pgmpy/tests/test_causal_discovery/test_HillClimbSearch.py]: why have you commented this? I think this test is being used for debugging purposes if I am not mistaken. It should be returning a dictionary with key-value pair consisting of the failed checks and its reason.

File: pgmpy/tests/test_causal_discovery/test_PC.py [hunk 2]
```diff
@@ -18,8 +21,8 @@
 
 def expected_failed_checks(estimator):
     return {
-        "check_fit_score_takes_y": "Causal discovery estimators do not take y parameter in score method.",
-        "check_n_features_in_after_fitting": "Failing for score method (not for fit) for unknown reason.",
+        # "check_fit_score_takes_y": "Causal discovery estimators do not take y parameter in score method.",
+        # "check_n_features_in_after_fitting": "Failing for score method (not for fit) for unknown reason.",
```
Comment (DARHWOLF) [inline pgmpy/tests/test_causal_discovery/test_PC.py]: same here

File: pgmpy/tests/test_causal_discovery/test_ExpertInLoop.py [hunk 3]
```diff
@@ -26,8 +27,8 @@ def make_estimator():
 
 def expected_failed_checks(estimator):
     return {
-        "check_fit_score_takes_y": "Causal discovery estimators do not take y parameter in score method.",
-        "check_n_features_in_after_fitting": "Failing for score method (not for fit) for unknown reason.",
+        # "check_fit_score_takes_y": "Causal discovery estimators do not take y parameter in score method.",
+        # "check_n_features_in_after_fitting": "Failing for score method (not for fit) for unknown reason.",
```
Comment (DARHWOLF) [inline pgmpy/tests/test_causal_discovery/test_ExpertInLoop.py]: refer comment

**Review by DARHWOLF — CHANGES_REQUESTED** [review]
- Please refer to my comment regarding the use of comments.
- may I request you to share how you debugged this issue.

**Comment by Famous077** [comment]
Hi @DARHWOLF , In this issue, Two specific sklearn checks were failing, In supervised learning, y label which is compare against predictions which was missing on score() function, due to which we are getting TypeError. another one is n_feature_in_ is set properly in def fit() function, but missing in def score() one. So, I add same line in def score() to prevent mismatches later. Because this functions tells sklearn how many features the model was trained on. lastly why I commented those line, these 2 checks failure are expected ones. So, whenever these checks get failed rather than produce warning those commented line were meant to be sent. So, I was verifying without those everything works perfectly or not, and as a result everything was working perfectly and all tests were passed.  Love to take feedback, if I made any mistakes.

**Comment by Famous077** [comment]
Hi @ankurankan @DARHWOLF  , Could you please review this PR whenever you have time.
